README.md
![build status](https://travis-ci.org/merklejerk/flex-ether.svg?branch=master)
![npm package](https://badge.fury.io/js/flex-ether.svg)

# flex-ether
A modern, flexible Ethereum library for sending ethereum transactions that:

- Requires minimal to *no* configuration to get going on all networks (no provider necessary).
- Can sign and send transactions from arbitrary wallets (private keys).
- Provides separate promises for transaction hashes, receipts, and confirmations.
- Automatically calculates gas and gas price for transactions in a configurable manner.
- Automatically resolves ENS addresses across all inputs.

## Installation
```bash
npm install flex-ether
# or
yarn install flex-ether
```

## Preview

```js
const FlexEther = require('flex-ether');
// A self-signing wallet key for transactions.
const PRIVATE_KEY = '0xb3734ec890893585330c71ece72afb05058192b6be47bee2b99714e6bb5696ab';

// Create instance on the mainnet.
let eth = new FlexEther();
// Send 100 wei from a self-signed wallet to an ENS address.
let tx = eth.transfer('ethereum.eth', '100', {key: PRIVATE_KEY});
// Wait for the transaction hash.
let transactionHash = await tx.txId;
// Wait for the receipt.
receipt = await tx.receipt;
// Wait for the receipt after 3 confirmations.
receipt = await tx.confirmed(3);
// Get the balance of an address at a certain block.
let balance = await eth.getBalance('0xf6fb5b73987d6d9a139e23bab97be6fc89e0dcd1', 412045);
// Estimate gas for a transaction, from a self-signed wallet.
let gas = eth.estimateGas('0xf6fb5b73987d6d9a139e23bab97be6fc89e0dcd1', '100',
   {key: PRIVATE_KEY});
```

## User Guide
- [Creating an instance](creating-an-instance)
- [Sending ether](#sending-ether)
- [Transaction promises](#transaction-promises)
- [Getting balances](#getting-balances)
- [Estimating gas](#estimating-gas)
- [ENS addresses](#ens-addresses)
- [Instance Properties](#instance-properties)

### Creating an instance
By default, the instance will create an [Infura](https://infura.io) provider to
talk to the main network. You can modify this behavior with the options
`network`, `infuraKey`, `provider`, or `providerURI`.

##### Full options
```js
eth = new FlexEther(
   // Options object. May be omitted.
   {
      // Network to use with Infura provider.
      // May be 'main', 'ropsten', 'rinkeby', or 'kovan'.
      // Defaults to 'main'
      network: String,
      // Infura Project ID, if not using a custom provider.
      infuraKey: String,
      // Connect to an existing provider at a URI
      // (e.g., http://localhost:8545 or https://mainnet.infura.io/v3/PROJECT_ID).
      // The 'net' option is required is using an IPC path.
      providerURI: String,
      // net instance, from require('net'), if using IPC path in providerURI
      net: Object,
      // Use a custom provider instance (e.g., web3.currentProvider for metamask).
      provider: Object,
      // Clamp transaction gas prices to this amount (in wei).
      // Defaults to 250 gwei.
      maxGasPrice: string,
      // Fractional bonus to apply to gas price when making transactions.
      // 0.01 = +1%. May be negative to under-price.
      // Defaults to -0.005.
      // Can be overridden in send/transfer calls.
      gasPriceBonus: Number,
      // Fractional bonus to apply to gas limit estimates when making transactions.
      // 0.01 = +1%. May be negative, but probably not a good idea.
      // Defaults to 0.66.
      // Can be overridden in send/transfer calls.
      gasBonus: Number,
      // ENS options.
      ens: {
          // Minimum number of seconds time to keep a resolved ENS name in cache.
          // Defaults to one hour.
          minTTL: Number,
          // Maximum number of seconds time to keep a resolved ENS name in cache.
          // Defaults to infinity.
          maxTTL: Number,
      }
   });
```

### Sending ether

Ether can be sent with the `transfer()` or the lower-level `send()` methods.

By default, transactions will be signed by the wallet associated with
the first account given by the provider. You can override the
caller by either passing the `from` or `key` option. The `from` option will
let the provider sign the transaction from an unlocked wallet, as usual.
But, the `key` option will *self-sign* the transaction with the private key
provided, allowing you to transact from any wallet you have the private keys
to.

Transactions return a [Transaction Promise Object](#transaction-promises), which
allow you to easily wait on transaction hashes,
[receipts](https://web3js.readthedocs.io/en/1.0/web3-eth.html#eth-gettransactionreceipt-return),
and confirmations.

##### Examples
```js
const FlexEther = require('flex-ether');
// A self-signing wallet key for transactions.
const PRIVATE_KEY = '0xb3734ec890893585330c71ece72afb05058192b6be47bee2b99714e6bb5696ab';
const eth = new FlexEther();

// Send 100 wei to an ENS address and wait for the receipt.
let receipt = await eth.transfer('ethereum.eth', '100');
/* Result: <Receipt Object> {
   transactionHash: '0x9eb3f89f8581e6c6df294344b538d44e265c226ae6e8ce6210df497cf2b54bd3',
   blockNumber: 3616104,
   gasUsed: 21000,
   ... etc.
}*/
// Send 100 wei to an address and wait for the transaction hash.
let txId = await eth.transfer(
   '0xf6fb5b73987d6d9a139e23bab97be6fc89e0dcd1', '100').txId;
// Send 100 wei to an address and wait for the receipt after 3 confirmations.
receipt = await eth.transfer(
   '0xf6fb5b73987d6d9a139e23bab97be6fc89e0dcd1', '100').confirmed(3);
// Send 100 wei from a wallet managed by the provider and wait for the receipt
receipt = await eth.transfer(
   '0xf6fb5b73987d6d9a139e23bab97be6fc89e0dcd1', '100',
   {from: '0x005B68A967D39c497074127871297b6728a1cfEd'});
// Send 100 wei from a wallet defined by a private key and wait for the receipt
receipt = await eth.transfer(
   '0xf6fb5b73987d6d9a139e23bab97be6fc89e0dcd1', '100', {key: PRIVATE_KEY});
// Same as above but with send().
receipt = await eth.send(
   '0xf6fb5b73987d6d9a139e23bab97be6fc89e0dcd1', {value: '100', key: PRIVATE_KEY});

```

##### Full options
```js
// Full method options.
await eth.transfer(
   // Recipient address. May be an ENS address.
   to: String,
   // Amount of ether to send, in weis.
   // Can be a base-10 string or hex-encoded number.
   amount: String,
   // Options.
   {
      // Address of wallet that will sign the transaction. May be an ENS address.
      // Must be unlocked by the provider.
      // Defaults to eth.getDefaultAccount().
      from: String,
      // Hex-encoded string private key.
      // Signs the transaction with this private key and sends it from the address
      // associated with it. Overrides 'from' option.
      key: String,
      // Gas price to use, as a hex or base-10 string, in wei.
      // If not specified, calculated from network gas price and bonus.
      gasPrice: String,
      // Execution gas limit.
      // If not specified, it will be estimated with bonus.
      gas: Number,
      // Bonus to apply to gas price calculations.
      // Should be a positive or negative Number, where 0.01 = +1%.
      // If omitted, `eth.gasPriceBonus` will be used.
      gasPriceBonus: undefined,
      // Bonus to apply to gas limit calculations.
      // Should be a positive or negative Number, where 0.01 = +1%.
      // If omitted, `eth.gasBonus` will be used.
      gasBonus: undefined
   });
```

### Transaction promises
`transfer()` and `send()` both return a Promise object that resolves
to the
[transaction receipt](https://web3js.readthedocs.io/en/1.0/web3-eth.html#eth-gettransactionreceipt-return),
once the transaction has been mined.

This Promise object also has the following properties:
- `txId`: a promise that resolves to the transaction hash when the transaction is
posted to the blockchain. This ususally comes much sooner than the receipt.
- `receipt`: a promise that resolves to the transaction receipt when the
transaction has been mined. Same as waiting on the parent object itself.
- `confirmed(count=1)` a function that returns a promise that resolves to the
transaction receipt after the transaction has been mined and `count` number of
confirmations have been seen, up to a maximum of 12 confirmations.

##### Example
```js
const FlexEther = require('flex-ether');
const eth = new FlexEther();

// Send 100 wei to an address and wait for the receipt.
let receipt = await eth.transfer(
   '0xf6fb5b73987d6d9a139e23bab97be6fc89e0dcd1', '100');
// Send 100 wei to an address and get the promise object.
let tx = eth.transfer(
   '0xf6fb5b73987d6d9a139e23bab97be6fc89e0dcd1', '100');
// Wait on the transaction hash.
let transactionHash = await tx.txId;
// Wait on the receipt. Equivalent to `await tx`
receipt = await tx.receipt;
// Wait on the receipt after 3 confirmations. Equivalent to `await tx`
receipt = await tx.confirmed(3);
```

### Getting balances
The `getBalance()` method queries the balance of an address.

You can also pass the block number at which to evaluate the balance. These
numbers can either be explicit block numbers or negative offsets from the last
block number, where `-1` is the last block, `-2` is the second to last block,
and so on.

##### Examples
```js
const FlexEther = require('flex-ether');
const eth = new FlexEther();

// Get the balance of an address at the current block.
let bal = await eth.getBalance('0xf6fb5b73987d6d9a139e23bab97be6fc89e0dcd1');
// Get the balance of an address at a specific block.
bal = await eth.getBalance('0xf6fb5b73987d6d9a139e23bab97be6fc89e0dcd1', 310319);
// Get the balance of an ENS address at the second to last block.
bal = await eth.getBalance('ethereum.eth', -2);

```

### Estimating gas
Sending ether from wallet to wallet generally costs exactly `21000` gas. However,
sending ether to a contract may actually cost more, as it might trigger the
execution of code. By default, the library will automatically compute and
allocate the gas needed before sending a transaction.

You can get the gas explicitly by calling `estimateGas()` with the same
parameters you would pass to `send()`.

##### Examples
```js
const FlexEther = require('flex-ether');
// A self-signing wallet key for transactions.
const PRIVATE_KEY = '0xb3734ec890893585330c71ece72afb05058192b6be47bee2b99714e6bb5696ab';
const eth = new FlexEther();

// Get the gas consumed by sending 100 wei to an address from the default account.
let gas = await eth.estimateGas('0xf6fb5b73987d6d9a139e23bab97be6fc89e0dcd1',
   {value: '100'});
// Get the gas consumed by sending 100 wei to an address from a wallet defined by
// a private key.
let gas = await eth.estimateGas('0xf6fb5b73987d6d9a139e23bab97be6fc89e0dcd1',
   {value: '100', key: PRIVATE_KEY});
```

### ENS addresses
Anywhere you can pass an address, you can instead pass an
[ENS address](http://docs.ens.domains/en/latest/introduction.html), such as
`'thisismyensaddress.eth'`. You can also call the `resolveAddress()` method
to resolve an address explicitly. If an ENS address cannot be resolved, an
exception will be raised (the promise will fail).

ENS is only available on the main, ropsten, and rinkeby networks.
The ENS address will also have to be set up with the ENS contract on the
respective network to properly resolve.

### Instance Properties
A contract instance exposes a few properties, most of which you are free to
change. Many of these can also be overridden in individual call options.

- `gasBonus (Number)` Gas limit estimate bonus for transactions, where `0.01 = +1%`. May be negative.
- `gasPriceBonus (Number)` Gas price bonus for transactions, where `0.01 = +1%`. May be negative.
- `async getTransactionCount(addr)` Get the nonce for an account.
- `async resolveBlockDirective(blockNum)` Resolve a block directive (e.g., `41204102` or `-2`) to a block number.
- `async getChainId()` Get the chain ID of the connected network.
- `async resolveAddress(addr)` Resolve an ENS address. If a regular address is passed, the checksummed version will be returned.
- `async getBlockNumber()` Get the current block number.
- `async getDefaultAccount()` Get the default account, set by the provider.

### Module Properties
The following module properties affect gas calculations for all instances:

- `MAX_GAS_PRICE` Maximum gas price for transactions. Defaults to `256` gwei.


CHANGELOG.md
## 1.4.0
- Remove dependency on `web3.js` and `ez-ens`.
- Make RPC calls ourselves.
- Embed our own ENS logic.


test/tests.js
'use strict'
const ganache = require('ganache-cli');
const FlexEther = require('../src/index');
const promisify = require('util').promisify;
const assert = require('assert');
const crypto = require('crypto');
const ethjs = require('ethereumjs-util');

describe('flex-contract', function() {
	let _ganache = null;
	let provider = null;
	let accounts = null;

	before(async function() {
		accounts = _.times(16, () => {
			const key = crypto.randomBytes(32);
			return {
				secretKey: key,
				address: ethjs.toChecksumAddress(
					'0x'+ethjs.privateToAddress(key).toString('hex')
				),
				balance: 100 + _.repeat('0', 18)
			};
		});
		provider = ganache.provider({
			accounts: accounts
		});
		// Suppress max listener warnings.
		provider.setMaxListeners(4096);
		provider.engine.setMaxListeners(4096);
	});

	it('can get balance', async function() {
		const eth = new FlexEther({provider: provider});
		assert.equal(await eth.getBalance(accounts[0].address), accounts[0].balance);
		assert.equal(await eth.getBalance(randomAddress()), '0');
	});

	it('can get balance at block number', async function() {
		const eth = new FlexEther({provider: provider});
		const block = await eth.getBlockNumber();
		const balance = await eth.getBalance(accounts[0].address);
		await eth.transfer(randomAddress(), 100);
		assert.equal(await eth.getBalance(accounts[0].address, block), balance);
	});

	it('can wait for receipt', async function() {
		const eth = new FlexEther({provider: provider});
		const to = randomAddress();
		const amount = _.random(1, 100);
		const tx = eth.transfer(to, amount);
		const receipt = await tx.receipt;
		assert.ok(receipt.transactionHash);
	});

	it('can wait for transaction hash', async function() {
		const eth = new FlexEther({provider: provider});
		const to = randomAddress();
		const amount = _.random(1, 100);
		const tx = eth.transfer(to, amount);
		const txId = await tx.txId;
		assert.ok(txId);
	});

	it('can wait for past confirmations', async function() {
		const eth = new FlexEther({provider: provider});
		const to = randomAddress();
		const amount = _.random(1, 100);
		const tx = eth.transfer(to, amount, {from: accounts[1].address});
		await tx;
		// Do some transactions to advance the block.
		for (let i = 0; i < 3; i++) {
			await eth.transfer(randomAddress(), _.random(1, 100));
		}
		const receipt = await tx.confirmed(3);
		assert.ok(receipt.transactionHash);
	});

	it('can wait for future confirmations', async function() {
		const eth = new FlexEther({provider: provider});
		const to = randomAddress();
		const amount = _.random(1, 100);
		const confirmed = (async () => {
			const tx = eth.transfer(to, amount, {from: accounts[1].address});
			return tx.confirmed(3);
		})();
		// Do some transactions to advance the block.
		for (let i = 0; i < 3; i++) {
			await eth.transfer(randomAddress(), _.random(1, 100));
		}
		const receipt = await confirmed;
		assert.ok(receipt.transactionHash);
	});

	it('can send ether from default account', async function() {
		const eth = new FlexEther({provider: provider});
		const from = await eth.getDefaultAccount();;
		const prevBalance = await eth.getBalance(from);
		const to = randomAddress();
		const amount = _.random(1, 100);
		const receipt = await eth.transfer(to, amount);
		assert.ok(receipt.transactionHash);
		assert.equal(await eth.getBalance(to), amount);
		assert.ok(await eth.getBalance(from) != prevBalance);
	});

	it('can send ether from provider account', async function() {
		const eth = new FlexEther({provider: provider});
		const from = accounts[1].address;
		const prevBalance = await eth.getBalance(from);
		const to = randomAddress();
		const amount = _.random(1, 100);
		const receipt = await eth.transfer(to, amount, {from: from});
		assert.ok(receipt.transactionHash);
		assert.equal(await eth.getBalance(to), amount);
		assert.ok(await eth.getBalance(from) != prevBalance);
	});

	it('can send ether with key', async function() {
		const eth = new FlexEther({provider: provider});
		const from = accounts[2].address;
		const key = accounts[2].secretKey;
		const to = randomAddress();
		const amount = _.random(1, 100);
		const prevBalance = await eth.getBalance(from);
		const receipt = await eth.transfer(to, amount, {key: key});
		assert.ok(receipt.transactionHash);
		assert.equal(await eth.getBalance(to), amount);
		assert.ok(await eth.getBalance(from) != prevBalance);
	});

	it('can estimate gas', async function() {
		const eth = new FlexEther({provider: provider});
		const to = randomAddress();
		const amount = _.random(1, 100);
		const gas = await eth.estimateGas(to, amount);
		assert.ok(gas >= 21000);
	});

	it('can call', async function() {
		const eth = new FlexEther({provider: provider});
		const to = randomAddress();
		const amount = _.random(1, 100);
		const r = await eth.call(to, amount);
		assert.equal(r, '0x');
	});

	it('can get balance from a prior block number', async function() {
		const eth = new FlexEther({provider: provider});
		const to = randomAddress();
		const blockNumber = await eth.getBlockNumber();
		const amount = _.random(100, 1000);
		const balance = await eth.getBalance(to, blockNumber);
		assert.equal(balance, '0');
	});

	it('can get balance from a prior block number by offset', async function() {
		const eth = new FlexEther({provider: provider});
		const to = randomAddress();
		const blockNumber = await eth.getBlockNumber();
		const amount = _.random(100, 1000);
		const balance = await eth.getBalance(to, -1);
		assert.equal(balance, '0');
	});

	it('can get gas price', async function() {
		const eth = new FlexEther({provider: provider});
		assert.ok(await eth.getGasPrice());
	});
});

function randomHex(size=32) {
	return '0x'+crypto.randomBytes(size).toString('hex');
}

function randomAddress() {
	return ethjs.toChecksumAddress(randomHex(20));
}


src/index.js
'use strict'
const _ = require('lodash');
const assert = require('assert');
const BigNumber = require('bignumber.js');
const cw3p = require('create-web3-provider');
const ethjstx = require('ethereumjs-tx');
const ethjs = require('ethereumjs-util');

const util = require('./util');
const Resolver = require('./resolver');
const RpcClient = require('./rpc-client');
const createTransactionPromise = require('./transaction-promise');

module.exports = class FlexEther {
	constructor(opts={}) {
		this.provider = opts.provider;
		if (!this.provider) {
			this.provider = cw3p({
				uri: opts.providerURI,
				net: opts.net
			});
		}
		this._rpc = new RpcClient(this.provider);
		this._resolver = new Resolver(this._rpc, opts.ens);
		this.gasBonus = _.isNumber(opts.gasBonus) ? opts.gasBonus : 0.66;
		this.gasPriceBonus = _.isNumber(opts.gasPriceBonus) ?
			opts.gasPriceBonus : -0.005;
		this.maxGasPrice = opts.maxGasPrice || new BigNumber('250e9').toString(10); // 250 gwei
	}

	async getChainId() {
		return this._rpc.getChainId();
	}

	async getDefaultAccount() {
		return this._rpc.getDefaultAccount();
	}

	async getTransactionCount(addr) {
		addr = await this._resolver.resolve(addr);
		return this._rpc.getTransactionCount(addr);
	}

	async getTransactionReceipt(txHash) {
		return this._rpc.getTransactionReceipt(await txHash);
	}

	async getBalance(addr, block='latest') {
		addr = await this._resolver.resolve(addr);
		return this._rpc.getBalance(addr, await this.resolveBlockDirective(block));
	}

	async getGasPrice() {
		return this._rpc.getGasPrice();
	}

	async getBlock(numberOrHash='latest') {
		return this._rpc.getBlock(await this.resolveBlockDirective(numberOrHash));
	}

	async getBlockNumber() {
		return this._rpc.getBlockNumber();
	}

	async _getGasPriceWithBonus(bonus) {
		bonus = (_.isNumber(bonus) ? bonus : this.gasPriceBonus) || 0;
		const price = new BigNumber(await this.getGasPrice()).times(1 + bonus);
		if (price.gt(this.maxGasPrice)) {
			price = new BigNumber(max);
		}
		return price.toString(10);
	}

	async estimateGas(to, opts={}) {
		const txOpts = await createTransactionOpts(this, to, opts);
		return await estimateGasRaw(this, txOpts, opts.gasBonus);
	}

	send(to, opts={}) {
		return createTransactionPromise(this, sendTx(this, to, opts));
	}

	transfer(to, amount, opts) {
		return createTransactionPromise(
			this,
			sendTx(
				this,
				to,
				_.assign({}, opts, { value: amount })
			)
		);
	}

	async call(to, opts={}) {
		return callTx(this, to, opts);
	}

	async getBlockNumber() {
		return this._rpc.getBlockNumber();
	}

	async resolveBlockDirective(block=-1) {
		if (block === 'latest' || block === 'pending' || block === 'earliest') {
			return block;
		}
		if (_.isNumber(block)) {
			if (block < 0) {
				let n = await this._rpc.getBlockNumber();
				n += block + 1;
				if (n <= 0) {
					throw Error(`Block number offset is too large: ${block}`);
				}
				return n;
			}
			return block;
		}
		throw new Error(`Invalid block directive: ${block}`);
	}

	async resolve(addr) {
		if (!addr)
			throw new Error('Invalid address.');
		return this._resolver.resolve(addr);
	}
};

async function getBlockGasLimit(inst) {
	while (true) {
		const lastBlock = await inst.getBlock();
		if (lastBlock != null)
			return lastBlock.gasLimit;
	}
}

async function estimateGasRaw(inst, txOpts, bonus) {
	txOpts = _.assign({}, txOpts, {
			gasPrice: 1,
			gasLimit: await getBlockGasLimit(inst),
		});
	bonus = (_.isNumber(bonus) ? bonus : inst.gasBonus) || 0;
	const gas = await inst._rpc.estimateGas(normalizeTxOpts(txOpts));
	return Math.ceil(gas * (1+bonus));
}

async function createTransactionOpts(inst, to, opts) {
	let from = undefined;
	if (opts.from === null) {
		// Explicitly leaving it undefined.
	} else if (_.isString(opts.from)) {
		from = await inst.resolve(opts.from);
	} else if (_.isNumber(opts.from)) {
		from = opts.from;
	} else if (opts.key)
		from = util.privateKeyToAddress(opts.key);
	else
		from = await inst.getDefaultAccount();
	to = to ? await inst.resolve(to) : undefined;
	return {
		gasPrice: opts.gasPrice,
		gas: opts.gasLimit || opts.gas,
		value: opts.value || 0,
		data: opts.data,
		to: to,
		from: from
	};
}

function normalizeTxOpts(opts) {
	const _opts = {};
	_opts.gasPrice = util.toHex(opts.gasPrice || 0);
	_opts.gas = util.toHex(opts.gasLimit || opts.gas || 0);
	_opts.value = util.toHex(opts.value || 0);
	if (!_.isNil(opts.nonce))
		_opts.nonce = parseInt(opts.nonce);
	if (opts.data && opts.data != '0x')
		_opts.data = util.toHex(opts.data);
	else
		_opts.data = '0x';
	if (_.isString(opts.to))
		_opts.to = ethjs.toChecksumAddress(opts.to);
	if (_.isString(opts.from))
		_opts.from = ethjs.toChecksumAddress(opts.from);
	return _opts;
}

async function callTx(inst, to, opts) {
	const block = _.isNil(opts.block) ?
		undefined : await inst.resolveBlockDirective(opts.block);
	const txOpts = await createTransactionOpts(inst, to, opts);
	_.defaults(txOpts, {
			gasPrice: 1,
			gasLimit: await getBlockGasLimit(inst)
		});
	if (!txOpts.to && (!txOpts.data || txOpts.data == '0x'))
		throw Error('Transaction has no destination.');
	return inst._rpc.call(normalizeTxOpts(txOpts), block);
}

async function sendTx(inst, to, opts) {
	const txOpts = await createTransactionOpts(inst, to, opts);
	if (!txOpts.from)
		throw Error('Cannot determine caller.');
	if (!txOpts.to && (!txOpts.data || txOpts.data == '0x'))
		throw Error('Transaction has no destination.');
	if (_.isNumber(opts.nonce))
		txOpts.nonce = opts.nonce;
	else
		txOpts.nonce = await inst.getTransactionCount(txOpts.from);
	if (!txOpts.gasPrice)
		txOpts.gasPrice = await inst.getGasPrice(opts.gasPriceBonus);
	if (!txOpts.gasLimit)
		txOpts.gasLimit = await estimateGasRaw(inst, txOpts, opts.gasBonus);
	if (!txOpts.chainId)
		txOpts.chainId = await inst._chainId;
	if (opts.key) {
		// Sign the TX ourselves.
		const tx = new ethjstx(normalizeTxOpts(txOpts));
		tx.sign(ethjs.toBuffer(opts.key));
		const serialized = util.toHex(tx.serialize());
		return inst._rpc.sendRawTransaction(serialized);
	}
	// Let the provider sign it.
	return inst._rpc.sendTransaction(normalizeTxOpts(txOpts));
}


src/transaction-promise.js
'use strict'

const CONFIRMATION_INTERVAL_MS = 4 * 1000;

function createTransactionPromise(flexEther, txHashPromise) {
    const confirmed = async function(minConfirmations=0) {
        const txHash = await txHashPromise;
        while (true) {
            const [
                receipt,
                confirmations
            ] = await getConfirmationsAndWait(flexEther, txHash);
            if (receipt && confirmations >= minConfirmations) {
                return receipt;
            }
        }
    };
    const txPromise = confirmed();
    txPromise.txHash = txPromise.transactionHash = txPromise.txId = txHashPromise;
    txPromise.confirmed = confirmed;
    txPromise.receipt = txPromise;;
    return txPromise;
}

async function getConfirmationsAndWait(flexEther, txHash) {
    // Need to keep fetching the receipt block number in case of a re-org.
    const [receipt, currentBlockNumber] = await Promise.all([
        await flexEther.getTransactionReceipt(txHash),
        await flexEther.getBlockNumber(),
    ]);
    if (receipt === null) {
        // No receipt available yet. Delay the response.
        return new Promise((accept, reject) => {
            setTimeout(() => accept([]), CONFIRMATION_INTERVAL_MS);
        });
    }
    return [receipt, currentBlockNumber - receipt.blockNumber];
};

module.exports = createTransactionPromise;


src/rpc-client.js
'use strict'
const _ = require('lodash');
const ethjs = require('ethereumjs-util');
const promisify = require('util').promisify;

module.exports = class RpcClient {
	constructor(provider) {
		this.provider = provider;
		this._chainId = null;
		this._nextRpcId = _.random(1, 2**64);
	}

	async getDefaultAccount() {
		return _.first(await this.getAccounts());
	}

	async getAccounts() {
		return this._send('eth_accounts');
	}

	async getTransactionCount(by, blockNumber='latest') {
		return toNumber(await this._send('eth_getTransactionCount',
			[
				asAddress(by),
				asBlockNumber(blockNumber)
			]
		));
	}

	async getBalance(by, blockNumber='latest') {
		return toUnsigned(await this._send('eth_getBalance',
			[
				asAddress(by),
				asBlockNumber(blockNumber)
			]
		));
	}

	async getGasPrice() {
		return toUnsigned(await this._send('eth_gasPrice'));
	}

	async getBlockNumber() {
		return toNumber(await this._send('eth_blockNumber'));
	}

	async getBlock(numberOrHash='latest') {
		let result;
		if (isHash(numberOrHash)) {
			result = await this._send('eth_getBlockByHash',
				[
					asHash(numberOrHash),
					false
				]
			);
		} else {
			result = await this._send('eth_getBlockByNumber',
				[
					asBlockNumber(numberOrHash),
					false
				]
			);
		}
		return normalizeBlock(result);
	}

	async estimateGas(tx) {
		return toNumber(await this._send('eth_estimateGas',
			[{
				to: asAddress(tx.to),
				from: !_.isNil(tx.from) ? asAddress(tx.from) : undefined,
				gas: !_.isNil(tx.gas) ? asHex(tx.gas) : undefined,
				gasPrice: !_.isNil(tx.gasPrice) ? asHex(tx.gasPrice) : undefined,
				value: !_.isNil(tx.value) ? asHex(tx.value) : undefined,
				data: !_.isNil(tx.data) ? asBytes(tx.data) : undefined,
			}]
		));
	}

	async call(tx, blockNumber='latest') {
		return await this._send('eth_call',
			[
				{
					to: asAddress(tx.to),
					from: !_.isNil(tx.from) ? asAddress(tx.from) : undefined,
					gas: !_.isNil(tx.gas) ? asHex(tx.gas) : undefined,
					gasPrice: !_.isNil(tx.gasPrice) ? asHex(tx.gasPrice) : undefined,
					value: !_.isNil(tx.value) ? asHex(tx.value) : undefined,
					data: !_.isNil(tx.data) ? asBytes(tx.data) : undefined,
				},
				asBlockNumber(blockNumber)
			]
		);
	}

	async sendTransaction(tx) {
		return this._send('eth_sendTransaction',
			[{
				to: asAddress(tx.to),
				from: !_.isNil(tx.from) ? asAddress(tx.from) : undefined,
				gas: !_.isNil(tx.gas) ? asHex(tx.gas) : undefined,
				gasPrice: !_.isNil(tx.gasPrice) ? asHex(tx.gasPrice) : undefined,
				value: !_.isNil(tx.value) ? asHex(tx.value) : undefined,
				data: !_.isNil(tx.data) ? asBytes(tx.data) : undefined,
			}]
		);
	}

	async sendRawTransaction(raw) {
		return this._send('eth_sendRawTransaction',
			[
				asBytes(raw)
			]
		);
	}

	async getTransactionReceipt(txHash) {
		const result = await this._send('eth_getTransactionReceipt',
			[
				asHash(txHash)
			]
		);
		return normalizeReceipt(result);
	}

	async getChainId() {
		if (typeof(this._chainId) === 'number') {
			return this._chainId;
		}
		return this._chainId = toNumber(await this._send('net_version'));
	}

	async _send(method, params=[]) {
		let sendPayload =
			this.provider.sendPayload ||
			this.provider.sendAsync ||
			this.provider.send;
		const numArgs = sendPayload.length;
		sendPayload = _.bind(sendPayload, this.provider);
		if (numArgs > 1) {
			sendPayload = promisify(sendPayload);
		}
		const { result } = await sendPayload({
			jsonrpc: "2.0",
			id: this._nextRpcId++,
			method: method,
			params: params
		});
		if (!result) {
			throw new RpcError(
				[
					`method=${JSON.stringify(method)}`,
					`params=${JSON.stringify(params)}`
				].join(', ')
			);
		}
		return result;
	}
}

const BLOCK_NUMBER_FIELDS = [
	'difficulty',
	'gasLimit',
	'gasUsed',
	'number',
	'size',
	'timestamp',
	'totalDifficulty'
];

function normalizeBlock(block) {
	for (const field of BLOCK_NUMBER_FIELDS) {
		block[field] = toNumber(block[field]);
	}
	return block;
}

const RECEIPT_NUMBER_FIELDS = [
	'blockNumber',
	'cumulativeGasUsed',
	'gasUsed',
	'status',
	'transactionIndex'
];

function normalizeReceipt(block) {
	for (const field of RECEIPT_NUMBER_FIELDS) {
		block[field] = toNumber(block[field]);
	}
	return block;
}

const HASH_REGEX = /^0x([0-9a-f]{2}){32}$/i;
const HEX_REGEX = /^0x[0-9a-f]*$/i;
const UNSIGNED_REGEX = /^[0-9]+$/;

function isHash(v) {
	return typeof(v) === 'string' && HASH_REGEX.test(v);
}

function asAddress(v) {
	if (!ethjs.isValidAddress(v)) {
		throw new InvalidAddressError(v);
	}
	return v;
}

function asHash(v) {
	if (!isHash(v)) {
		throw new InvalidHashError(v);
	}
	return v;
}

function asBlockNumber(v) {
	if (_.isNil(v)) {
		return 'latest';
	}
	if (v === 'latest' || v === 'pending' || v === 'earliest') {
		return v;
	}
	try {
		return asHex(v);
	} catch (err) {
		throw new InvalidBlockNumberError(v);
	}
}

function asBytes(v) {
	if (v == '0x') {
		return v;
	}
	if (!HEX_REGEX.test(v)) {
		throw new InvalidBytesError(v);
	}
	return ethjs.bufferToHex(ethjs.toBuffer(v));
}

function asHex(v) {
	const bn = toBN(v);
	return '0x' + bn.toString(16);
}

function toNumber(v) {
	return _.toNumber(v);
}

function toUnsigned(v) {
	return toBN(v).toString(10);
}

function toBN(v) {
	if (typeof(v) === 'number') {
		return new ethjs.BN(v);
	}
	if (typeof(v) === 'string') {
		if (HEX_REGEX.test(v)) {
			return new ethjs.BN(v.substr(2), 16);
		}
		if (UNSIGNED_REGEX.test(v)) {
			return new ethjs.BN(v);
		}
	}
	throw new InvalidUnsignedError(v);
}

class RpcError extends Error {
	constructor(msg) {
		super(msg);
		this.name = this.constructor.name;
	}
};

class InvalidAddressError extends Error {
	constructor(v) {
		super(`Invalid address: ${JSON.stringify(v)}`);
		this.name = this.constructor.name;
	}
};

class InvalidHashError extends Error {
	constructor(v) {
		super(`Invalid hash: ${JSON.stringify(v)}`);
		this.name = this.constructor.name;
	}
};

class InvalidBlockNumberError extends Error {
	constructor(v) {
		super(`Invalid block number: ${JSON.stringify(v)}`);
		this.name = this.constructor.name;
	}
};

class InvalidNumberError extends Error {
	constructor(v) {
		super(`Invalid number: ${JSON.stringify(v)}`);
		this.name = this.constructor.name;
	}
};

class InvalidUnsignedError extends Error {
	constructor(v) {
		super(`Invalid unsigned number: ${JSON.stringify(v)}`);
		this.name = this.constructor.name;
	}
};

class InvalidBytesError extends Error {
	constructor(v) {
		super(`Invalid bytes: ${JSON.stringify(v)}`);
		this.name = this.constructor.name;
	}
};

module.exports.RpcError = RpcError;
module.exports.InvalidAddressError = InvalidAddressError;
module.exports.InvalidHashError = InvalidHashError;
module.exports.InvalidBlockNumberError = InvalidBlockNumberError;
module.exports.InvalidNumberError = InvalidNumberError;
module.exports.InvalidUnsignedError = InvalidUnsignedError;
